frontend/webapp/utils/radius/forms.py

```diff
@@ -0,0 +1,126 @@
+    remember = BooleanField(get_form_field_label("remember_me"))
+    submit = SubmitField(get_form_field_label("login"))
+
+    def __init__(self, *args, **kwargs):
+        """Init new RADIUS login form."""
+        super(RADIUSLoginForm, self).__init__(*args, **kwargs)
+        self._args = args
+        self._kwargs = kwargs
+        if not self.next.data:
+            self.next.data = request.args.get("next", "")
+        self.remember.default = config_value("DEFAULT_REMEMBER_ME")
+
+    def validate(self):
+        """Validate."""
+
+        log = current_app.artemis_logger
+
+        log.info("RADIUS authenticate() using username %s" % self.email.data)
+
+        if not super(RADIUSLoginForm, self).validate():
+            log.warn("super() validate was false!")
+            return False
+
+        if self.email.data.strip() == "":
+            self.email.errors.append(get_message("USERID_NOT_PROVIDED")[0])
+            log.warn("userid was not provided")
+            return False
+
+        if self.password.data.strip() == "":
+            log.warn("password was not provided")
+            self.password.errors.append(get_message("PASSWORD_NOT_PROVIDED")[0])
+            return False
+
+        try:
+            admin_user = _datastore.get_user(1)
+            if self.email.data == admin_user.username:
+                log.info("Login using Super-user login")
+                return self._try_local_auth()
+
+            auth_result, role = _datastore.authenticate(
+                self.email.data, self.password.data
+            )
+            if auth_result is None:
+                self.password.errors.append("No response from RADIUS")
+                log.info("RADIUS authenticate() returned None")
+                return False
+            if not auth_result:
+                self.password.errors.append(get_message("INVALID_PASSWORD")[0])
+                log.info("RADIUS authenticate() returned False")
+                return False
+            log.info("RADIUS authenticate() returned True. Assigning role %s" % role)
+            username = self.email.data
+            self.user = _datastore.find_user(username=username)
+            if self.user:
+                self.user.password = None
+                self.user.email = username
+            else:
+                self.user = _datastore.create_user(
+                    username=username, email=username, password=None, active=True
+                )
+            self._set_role(self.user, _datastore.find_role(role))
+            _datastore.commit()
+        except Exception:
+            self.password.errors.append(
+                "Internal error. Contact developer and/or check the logs."
+            )
+            log.exception("Unexpected error while handling RADIUS form")
+            return False
+
+        return True
+
+    # because the user manipulation is broken i.e. it can lead to multiple entries in roles_users table (perhaps lack of
```

Comment:
@slowr 
I have discovered a bug in user manipulation code (tested on master).
To reproduce:
* create a user with user role
* delete this user from the UI
* create the same user (same username) with . create_user() method
* assign this user admin role using .add_role_to_user() method

now check the database; it will contain 2 entries in roles_users table for that user.

I don't know where should it be fixed. The natural place would be to add primary key constraint in roles_users table but I am not competenent enough to do that.
Other approach would be to remove user roles before removing him from the ui (actions_view.py)

Pls. can you comment on that?

Reply:
@pputerla I merged a fix for this. Can you verify that in your example there is no issue anymore?